components/in_memory_engine/src/background.rs

```diff
@@ -1330,10 +1332,37 @@ impl RunnableWithTimer for BackgroundRunner {
         IN_MEMORY_ENGINE_MEMORY_USAGE.set(mem_usage as i64);
 
         let mut count_by_state = [0; RegionState::COUNT];
+        let mut oldest_safe_point = u64::MAX;
+        let mut newest_safe_point = u64::MIN;
         {
             let regions_map = self.core.engine.region_manager().regions_map.read();
             for m in regions_map.regions().values() {
                 count_by_state[m.get_state() as usize] += 1;
+                if m.get_state() == RegionState::Active && m.safe_point() != 0 {
+                    oldest_safe_point = u64::min(oldest_safe_point, m.safe_point());
+                    newest_safe_point = u64::max(newest_safe_point, m.safe_point());
+                }
+            }
+        }
+
+        if oldest_safe_point != u64::MAX {
+            IN_MEMORY_ENGINE_OLDEST_SAFE_POINT.set(oldest_safe_point as i64);
+            IN_MEMORY_ENGINE_NEWEST_SAFE_POINT.set(newest_safe_point as i64);
+            if let Ok(tikv_safe_point) = block_on(self.pd_client.get_gc_safe_point()) {
+                if tikv_safe_point < oldest_safe_point {
+                    let gap = TimeStamp::new(oldest_safe_point - tikv_safe_point).physical();
+                    // If gap is too larger (more than a year), it means tikv safe point is not
+                    // initialized, so we does not update the metrics now.
+                    if gap < Duration::from_secs(365 * 24 * 3600).as_millis() as u64 {
+                        SAFE_POINT_GAP.set((oldest_safe_point - tikv_safe_point) as i64);
```

Comment:
It can be set a negative value for the case that tikv gc safepoint < ime gc safepoint.

Reply:
We have `if tikv_safe_point < oldest_safe_point {` above.

Reply:
how can be a negative value?